### PR TITLE
Alternative instance with less backtracking

### PR DESCRIPTION
**Pull request by considerate**

### CODE REVIEW

**Comment by cdepillabout** [comment]
@rprije I don't know if you heard about `aeson-commit`, but I'm under the understanding that it is a library from Jonas and Viktor that has a different JSON (yaml?) parsing strategy than aeson.  It doesn't backtrack as aggressively, which leads to much better error messages.

I just got this message from Viktor on Slack:

> I've now verified with some tests in aeson-commit that with the new Alternative instance we get the correct backtracking behavior. What I also managed to verify is that we **don't** get the correct JSONPath in the error messages. The path is discarded every time and I can't seem to figure out how to solve it.

I think Viktor has a failing(?) test case for this:

https://github.com/xc-jp/aeson-commit/pull/1/files#diff-d4282d58621e6f1530118b0876dc1caaR56

Or maybe it just should be failing, and it is not?

File: src/Data/Aeson/Commit.hs [hunk 1]
```diff
@@ -21,25 +22,36 @@ module Data.Aeson.Commit
...
 
 -- | Create a commit parser that doesn't backtrack if the first parser parses
 -- successfully.
 commit :: (Value -> Parser x) -> (x -> Parser y) -> Commit y
-commit f g = Commit . ReaderT $ MaybeT . pure . parseMaybe f >=> lift . g
+commit f g = Commit . ReaderT $ \v -> MaybeT $ do
+    mx <- optional (f v)
+    case mx of
+      Nothing -> pure Nothing
+      Just x  -> Just <$> g x
```
Comment (jonascarpay) [inline src/Data/Aeson/Commit.hs:54]: isn't this `fmap g <$> optional (f v)`?

**Review by jonascarpay — APPROVED** [review]

**Comment by rprije** [comment]
> @rprije I don't know if you heard about `aeson-commit`, but I'm under the understanding that it is a library from Jonas and Viktor that has a different JSON (yaml?) parsing strategy than aeson. It doesn't backtrack as aggressively, which leads to much better error messages.
> 
> I just got this message from Viktor on Slack:
> 
> > I've now verified with some tests in aeson-commit that with the new Alternative instance we get the correct backtracking behavior. What I also managed to verify is that we **don't** get the correct JSONPath in the error messages. The path is discarded every time and I can't seem to figure out how to solve it.
> 
> I think Viktor has a failing(?) test case for this:
> 
> https://github.com/xc-jp/aeson-commit/pull/1/files#diff-d4282d58621e6f1530118b0876dc1caaR56
> 
> Or maybe it just should be failing, and it is not?

I had a look at this. I don't think the test case exercises anything in Aeson that produces a path. When I convert it to ordinary Aeson, no path is given in the output. To get a path, I must decode a string. Here's some test cases I used:

```
nestedFail :: IO ()
nestedFail = parseCommit parser value `shouldBe` Left "Error in $.a.c: blaargh"
  where
  value = object ["a" .= object ["c" .= ("d" :: String)]]
  parser :: Commit String
  parser = commit (parseKey "a") (runCommit inner)
  inner :: Commit String
  inner = commit (parseKey "c") (\v -> fromValue v (fail "blaargh"))

ordinaryNestedFail :: IO ()
ordinaryNestedFail = parseEither parser value `shouldBe` Left "Error in $.a.c: blaargh"
  where
  value = object ["a" .= object ["c" .= ("d" :: String)]]
  parser :: Value -> Parser String
  parser v = do
    a <- parseKey "a" v
    c <- parseKey "c" a
    fromValue c $ fail "blaargh"

stringNestedFail :: IO ()
stringNestedFail =
  (eitherDecode "{ \"a\" : [[1,2],[null,3]] }" :: Either String (Map String [[Int]]))
    `shouldBe` Left "Error in $.a[1][0]: parsing Int failed, expected Number, but encountered Null"
```

And the output:

```
est suite tasty: RUNNING...
Commit
  Backtracking
    commits:              OK
    backtracks:           OK
    no match:             OK
    nested fail:          FAIL
      expected: Left "Error in $.a.c: blaargh"
       but got: Left "Error in $: blaargh"
    ordinary nested fail: FAIL
      expected: Left "Error in $.a.c: blaargh"
       but got: Left "Error in $: blaargh"
    string nested fail:   OK

2 out of 6 tests failed (0.00s)
Test suite tasty: FAIL
```

Note that only `stringNestedFail` making use of the bytestring decoder actually produces an expected json path in the error.

This seems like a bug in Aeson, since that `$` clearly indicates that it's running Aeson's `formatError` but it seems to be being passed an empty path.

I'll need to dig further to ascertain whether the problem is that we're constructing the JSON object by hand, or something to do with constructing the parser by hand.

Maybe it would help to have an example of code which produces a path in vanilla Aeson for which we want  to observe similar behaviour, minus backtracking, in Aeson-Commit.

**Comment by rprije** [comment]
Updating this PR: we established on the call that this lack of a path in vanilla Aeson is a symptom of exactly what we're trying to fix. So we should not expect aeson-commit to behave the same way.

**Comment by considerate** [comment]
@rprije Thank you for the test case that verifies the error behavior. Do you get the same behavior if you first decode to `Value` and then bind that to a `Parser (Map String [[Int]])`?

**Comment by rprije** [comment]
> @rprije Thank you for the test case that verifies the error behavior. Do you get the same behavior if you first decode to `Value` and then bind that to a `Parser (Map String [[Int]])`?

Here are some test cases:

```
stringNestedFail :: IO ()
stringNestedFail =
  (eitherDecode "{ \"a\" : { \"c\" : null } }" :: Either String (Map String (Map String Int)))
    `shouldBe` Left "Error in $.a.c: parsing Int failed, expected Number, but encountered Null"

commitStringNestedFail :: IO ()
commitStringNestedFail =
  parseCommit parser value
    `shouldBe` Left "Error in $.a.c: parsing Int failed, expected Number, but encountered Null"
  where
    value :: Value
    value =
      case eitherDecode "{ \"a\" : { \"c\" : null } }" of
        Left s  -> error "Decode Fail"
        Right v -> v
    parser :: Commit Int
    parser = commit (parseKey "a") (runCommit inner)
    inner :: Commit Int
    inner = commit (parseKey "c") parseJSON

ordinaryStringNestedFail :: IO ()
ordinaryStringNestedFail =
  parseEither parser value
    `shouldBe` Left "Error in $.a.c: parsing Int failed, expected Number, but encountered Null"
  where
    value :: Value
    value =
      case eitherDecode "{ \"a\" : { \"c\" : null } }" of
        Left s  -> error "Decode Fail"
        Right v -> v
    parser :: Value -> Parser (Map String (Map String Int))
    parser = withObject "a" $ \o -> do
      a <- o .: "a"
      c <- a .: "c"
      parseJSON c

ordinaryStringNestedFail2 :: IO ()
ordinaryStringNestedFail2 =
  parseEither parser value
    `shouldBe` Left "Error in $.a.c: parsing Int failed, expected Number, but encountered Null"
  where
    value :: Value
    value =
      case eitherDecode "{ \"a\" : { \"c\" : null } }" of
        Left s  -> error "Decode Fail"
        Right v -> v
    parser :: Value -> Parser (Map String (Map String Int))
    parser = parseJSON
```

This results in:

```
    string nested fail:            OK
    commit string nested fail:     FAIL
      expected: Left "Error in $.a.c: parsing Int failed, expected Number, but encountered Null"
       but got: Left "Error in $: parsing Int failed, expected Number, but encountered Null"
    ordinary string nested fail:   FAIL
      expected: Left "Error in $.a.c: parsing Int failed, expected Number, but encountered Null"
       but got: Left "Error in $: parsing Map failed, expected Object, but encountered Null"
    ordinary string nested fail 2: OK
```

So the problem isn't the decoding of the string. The problem appears to be a difference between parsing a `Value` then drilling into it to get to some inner `Value` which fails, and parsing the entire `Value` as one Haskell data type via recursive calls to something like `parseJSON`.

It's not yet really clear to me why. `.:` is certainly built on top of `<?>` so it seems like it should track path components.

**Comment by rprije** [comment]
For what it's worth, This worked:

```
commitStringNestedFail2 :: IO ()
commitStringNestedFail2 =
  parseCommit parser value
    `shouldBe` Left "Error in $.a.c: parsing Int failed, expected Number, but encountered Null"
  where
    value :: Value
    value =
      case eitherDecode "{ \"a\" : { \"c\" : null } }" of
        Left s  -> error "Decode Fail"
        Right v -> v
    parser :: Commit (Map String (Map String Int))
    parser = commit pure parseJSON
```

```
    commit string nested fail 2:   OK
```

Swapping the arguments to `commit` caused it to fail with a different error that loses information though:

```
commitStringNestedFail2 :: IO ()
commitStringNestedFail2 =
  parseCommit parser value
    `shouldBe` Left "Error in $.a.c: parsing Int failed, expected Number, but encountered Null"
  where
    value :: Value
    value =
      case eitherDecode "{ \"a\" : { \"c\" : null } }" of
        Left s  -> error "Decode Fail"
        Right v -> v
    parser :: Commit (Map String (Map String Int))
    parser = commit parseJSON pure
```

```
    commit string nested fail 2:   FAIL
      expected: Left "Error in $.a.c: parsing Int failed, expected Number, but encountered Null"
       but got: Left "Error in $: No parser matches value Object (fromList [(\"a\",Object (fromList [(\"c\",Null)]))])"
```

**Comment by rprije** [comment]
Small correction to the `ordinaryStringNestedFail` test case above: The `parser` type should be `Value -> Parser Int`

`Parser` is basically a reader monad. It can't preserve state changes to the path across bind.

The `Monad` instance for `Parser` passes the same path to both the left and right sides of the bind:

```
newtype Parser a = Parser {
      runParser :: forall f r.
                   JSONPath
                -> Failure f r
                -> Success a f r
                -> f r
    }

instance Monad.Monad Parser where
    m >>= g = Parser $ \path kf ks -> let ks' a = runParser (g a) path kf ks
                                       in runParser m path kf ks'
    {-# INLINE (>>=) #-}
    return = pure
    {-# INLINE return #-}
```

So `m` and `g` both receive the same path. `m` modifies its path, but it never reaches `g`.

The path is only built up when a single parser is taken and modified through non-monadic operations such as `<?>` and `<$>`. `Traversal` instances are built upon `Applicative` and should be fine. Notice how the failing examples above try to extract a nested component monadically, while the working examples parse and preserve the whole structure of the JSON object.

This example using traversal works and is basically what the generics derived FromJSON instances do:

```
ordinaryStringNestedFail3 :: IO ()
ordinaryStringNestedFail3 =
  parseEither parser value
    `shouldBe` Left "Error in $.a.c: parsing Int failed, expected Number, but encountered Null"
  where
    value :: Value
    value =
      case eitherDecode "{ \"a\" : { \"c\" : null } }" of
        Left s  -> error "Decode Fail"
        Right v -> v
    parser :: Value -> Parser (HashMap Text (HashMap Text Int))
    parser = withObject "a" $
      HashMap.traverseWithKey (\k v -> inner v <?> Key k)
    inner :: Value -> Parser (HashMap Text Int)
    inner = withObject "a" $
      HashMap.traverseWithKey (\k v -> parseJSON v <?> Key k)
```

```
    ordinary string nested fail 3: OK
```

**Comment by rprije** [comment]
`Parser` is basically a reader monad.

**Pull request merged**